docs/source/api/notebooks/Notebook_4_Customized_Models.ipynb

```diff
@@ -0,0 +1,456 @@
+    "        self.cr_attn_weights = nn.Parameter(th.Tensor(out_dim, 1))\n",
+    "        nn.init.normal_(self.cr_attn_weights)\n",
+    "\n",
+    "        # bias\n",
+    "        if bias:\n",
+    "            self.h_bias = nn.Parameter(th.Tensor(out_dim))\n",
+    "            nn.init.zeros_(self.h_bias)\n",
+    "\n",
+    "        # weight for self loop\n",
+    "        if self.self_loop:\n",
+    "            self.loop_weight = nn.Parameter(th.Tensor(in_dim, out_dim))\n",
+    "            nn.init.xavier_uniform_(self.loop_weight, gain=nn.init.calculate_gain('relu'))\n",
+    "\n",
+    "        # dropout\n",
+    "        self.dropout = nn.Dropout(dropout)\n",
+    "\n",
+    "        # normalization for each node type\n",
+    "        ntypes = set()\n",
+    "        for rel in rel_names:\n",
+    "            ntypes.add(rel[0])\n",
+    "            ntypes.add(rel[2])\n",
+    "\n",
+    "        if norm == \"batch\":\n",
+    "            self.norm = nn.ParameterDict({ntype:nn.BatchNorm1d(out_dim) for ntype in ntypes})\n",
+    "        elif norm == \"layer\":\n",
+    "            self.norm = nn.ParameterDict({ntype:nn.LayerNorm(out_dim) for ntype in ntypes})\n",
+    "        else:\n",
+    "            self.norm = None\n",
+    "\n",
+    "    def forward(self, g, inputs):\n",
+    "        \"\"\"\n",
+    "        g: DGL.block\n",
+    "            A DGL block\n",
+    "        inputs : dict[str, torch.Tensor]\n",
+    "            Node feature for each node type.\n",
+    "\n",
+    "        Returns\n",
+    "        -------\n",
+    "        dict[str, torch.Tensor]\n",
+    "            New node features for each node type.\n",
+    "        \"\"\"\n",
+    "        g = g.local_var()\n",
+    "\n",
+    "        # loop each edge type to fulfill GAT computation within each edge type\n",
+    "        for src_type, e_type, dst_type in g.canonical_etypes:\n",
+    "\n",
+    "            # extract subgraph of each edge type\n",
+    "            sub_graph = g[src_type, e_type, dst_type]\n",
+    "\n",
+    "            # check if no edges exist for this edge type\n",
+    "            if sub_graph.num_edges() == 0:\n",
+    "                continue\n",
+    "\n",
+    "            # extract source and destination node features\n",
+    "            src_feat = inputs[src_type]\n",
+    "            dst_feat = inputs[dst_type][ :sub_graph.num_dst_nodes()]\n",
+    "\n",
+    "            # GAT in one relation type\n",
+    "            agg = self.rel_gats[str((src_type, e_type, dst_type))](sub_graph, (src_feat, dst_feat))\n",
+    "            agg = agg.view(agg.shape[0], -1)\n",
+    "\n",
+    "            # store aggregations in destination nodes\n",
+    "            sub_graph.dstdata['agg_' + str((src_type, e_type, dst_type))] = self.leaky_relu(agg)\n",
+    "\n",
+    "        h = {}\n",
+    "        for n_type in g.dsttypes:\n",
+    "            if g.num_dst_nodes(n_type) == 0:\n",
```

Comment:
If a node does not have neighbors, the code will crash. (agg_list will be empty)

Reply:
Good catch. Add a check and raise error for no in-degree nodes.